### PR TITLE
ValueTuple to sequence value conversion

### PR DESCRIPTION
**Pull request by nblumhardt**
This PR for #965 converts tuples like `(1, 2, 3)` into arrays like `[1, 2, 3]` for consumption by sinks.

I don't think tuples are important enough, currently, to justify the effort of including another `LogEventPropertyValue` type to represent them. Arrays are a bit easier on the eyes than structures, and we don't lose any information since tuple member names aren't available dynamically.

The `@` operator is not required to capture a tuple. E.g.:

```csharp
Log.Information("{T}", (1, 2));
```

is fine. Specifying `@` will trigger deep serialization in the case of tuple members that are complex types.

The PR is a little ham-fisted (the change doesn't fit in as an `IScalarConversionPolicy`, unfortunately, so special treatment is needed). Only tuples up to length 7 are supported.

Since tuples are here to stay, and supported everywhere Serilog runs, I've introduced the _System.ValueTuple_ dependency for the current targets. A .NET 4.7-specific target would avoid it, as will .NET Core 2.0. We can add this fresh round of targets once core 2.0 ships, I think (4.7 is a pain right now as it requires a targeting pack).

### CODE REVIEW

File: src/Serilog/Parameters/PropertyValueConverter.cs [hunk 1]
```diff
@@ -190,6 +190,33 @@ LogEventPropertyValue CreatePropertyValue(object value, Destructuring destructur
                     enumerable.Cast<object>().Take(_maximumCollectionCount).Select(o => limiter.CreatePropertyValue(o, destructuring)));
             }
 
+            if (value is IStructuralEquatable)
```
Comment (nblumhardt) [inline src/Serilog/Parameters/PropertyValueConverter.cs]: Attempting to keep the cost of the test down.

File: src/Serilog/Parameters/PropertyValueConverter.cs [hunk 2]
```diff
@@ -190,6 +190,33 @@ LogEventPropertyValue CreatePropertyValue(object value, Destructuring destructur
...
+                if (type.IsConstructedGenericType)
+                {
+                    var definition = type.GetGenericTypeDefinition();
+                    if (definition == typeof(ValueTuple<>) || definition == typeof(ValueTuple<,>) ||
+                        definition == typeof(ValueTuple<,,>) || definition == typeof(ValueTuple<,,,>) ||
+                        definition == typeof(ValueTuple<,,,,>) || definition == typeof(ValueTuple<,,,,,>) ||
+                        definition == typeof(ValueTuple<,,,,,,>)) // Ignore the 8+ value case for now.
+                    {
+                        var elements = new List<LogEventPropertyValue>();
+                        foreach (var field in type.GetTypeInfo().DeclaredFields)
```
Comment (nblumhardt) [inline src/Serilog/Parameters/PropertyValueConverter.cs]: Assumes fields are ordered (I think this is safe).

**Comment by nblumhardt** [comment]
Ready to go, if CI likes me :-)

File: src/Serilog/Serilog.csproj [hunk 3]
```diff
@@ -23,25 +23,44 @@
   <ItemGroup Condition=" '$(TargetFramework)' == 'net45' ">
     <Reference Include="System" />
     <Reference Include="Microsoft.CSharp" />
+    <PackageReference Include="System.ValueTuple" Version="4.3.1" />
```
Comment (skomis-mm) [inline src/Serilog/Serilog.csproj]: It's a pity. Why you didn't considered the fourth option (along with the first)? 
> 4.Support it everywhere, using reflection only (no direct dependency on  ValueTuple  itself)

Is that because of some performance concerns? I believe you could check `type.Name` to constants (`ValueTyple'1`, `ValueTuple'2` etc) like you did with `definition ==` comparisons

File: src/Serilog/Serilog.csproj [hunk 4]
```diff
@@ -23,25 +23,44 @@
...
-    <PackageReference Include="System.Dynamic.Runtime" Version="4.0.11" />
-    <PackageReference Include="System.Globalization" Version="4.0.11" />
-    <PackageReference Include="System.Linq" Version="4.1.0" />
-    <PackageReference Include="System.Reflection" Version="4.1.0" />
-    <PackageReference Include="System.Reflection.Extensions" Version="4.0.1" />
-    <PackageReference Include="System.Runtime" Version="4.1.0" />
-    <PackageReference Include="System.Runtime.Extensions" Version="4.1.0" />
-    <PackageReference Include="System.Text.RegularExpressions" Version="4.1.0" />
-    <PackageReference Include="System.Threading" Version="4.0.11" />
+    <PackageReference Include="Microsoft.CSharp" Version="4.3.0" />
```
Comment (khellang) [inline src/Serilog/Serilog.csproj]: The `netstandard1.x` targets should probably be updated to just depend on `NETStandard.Library` anyway, TBH.

File: src/Serilog/Serilog.csproj (hunk 3 above)
Comment (nblumhardt) [inline src/Serilog/Serilog.csproj]: Thanks for the suggestion. I initially discounted this because it seemed a bit awkward, given tuples are a feature of C# that's here to stay. On second thought, it's not much worse to do the type name check than to do what we have here, so I'll switch it back over and `#if VALUETUPLE` conditionally-include the `Type`-based version 👍

File: src/Serilog/Serilog.csproj (hunk 4 above)
Comment (nblumhardt) [inline src/Serilog/Serilog.csproj]: Thanks @khellang - you're probably right here. I guess the dust has settled enough for us to embrace the new order of things :-). Perhaps we make the change when .NET Core 2.0 comes along? I am tempted to leave it alone in this PR, just to avoid any more chasing of corner cases....

**Comment by nblumhardt** [comment]
Alrighty! Avoided the package dependency and thus any dependency changes. Ready to go 🚀

File: test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs [hunk 5]
```diff
@@ -253,6 +255,55 @@ public void CSharpAnonymousTypesAreRecognizedWhenDestructuring()
...
+                (1, 2, 3, 4, 5, 6),
+                (1, 2, 3, 4, 5, 6, 7)
+            };
+
+            foreach (var t in tuples)
+                Assert.IsType<SequenceValue>(_converter.CreatePropertyValue(t));
+        }
+
+        [Fact]
+        public void EightPlusValueTupleElementsAreIgnored()
```
Comment (merbla) [inline test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs]: Nit `EightPlusValueTupleElementsAreIgnored` -> `EightPlusValueTupleElementsAreIgnoredWhenDestructuring`

Why the arbitrary 8?

File: test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs [hunk 6]
```diff
@@ -253,6 +255,55 @@ public void CSharpAnonymousTypesAreRecognizedWhenDestructuring()
...
+            Assert.Equal(new ScalarValue("A"), sequenceValue.Elements[1]);
+            var nested = Assert.IsType<SequenceValue>(sequenceValue.Elements[2]);
+            Assert.Equal(1, nested.Elements.Count);
+            Assert.Equal(new ScalarValue("B"), nested.Elements[0]);
+        }
+
+        [Fact]
+        public void AllTupleLengthsUpToSevenAreSupported()
+        {
+            var tuples = new object[]
```
Comment (merbla) [inline test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs]: Is an empty array test worthwhile?

Comment (merbla) [inline test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs]: Should we migrate to `InlineData` for use cases?

Comment (nblumhardt) [inline test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs]: Tried, but `ValueTuple` isn't regarded by csc as a compile-time constant, so no joy :-(

File: test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs (hunk 5 above)
Comment (nblumhardt) [inline test/Serilog.Tests/Parameters/PropertyValueConverterTests.cs]: 👍 probably should be `WhenCapturing`, but good point.

The limitation is in how `ValueTuple` works, up to seven elements can be represented directly in one tuple, but to go 8+ they need to be nested. No major issue with implementing it down the road.
